### PR TITLE
Initial support for i.MX 93 machines

### PR DESCRIPTION
**Pull request by angolini**
I have tested the PR on `imx93-11x11-lpddr4x-evk` and it boots.

While it's been reviewed, I'm working to fix `linux-fslc-imx` because today it does not boot for this board

### CODE REVIEW

File: conf/machine/include/imx-base.inc [hunk 1]
```diff
@@ -164,6 +164,7 @@ DEFAULTTUNE:mx8m-generic-bsp   ?= "cortexa53-crypto"
 DEFAULTTUNE:mx8qm-generic-bsp  ?= "cortexa72-cortexa53-crypto"
 DEFAULTTUNE:mx8x-generic-bsp   ?= "cortexa35-crypto"
 DEFAULTTUNE:mx8ulp-generic-bsp ?= "cortexa35-crypto"
+DEFAULTTUNE:mx9-generic-bsp    ?= "cortexa35-crypto"
```
Comment (zandrey) [inline conf/machine/include/imx-base.inc:167]: Not sure this tune is correct here.

IIRC, `imx93` has _A55_ and it is closer to _A53_ than to _A35_.

There is no `DEFAULTTUNE` for _A55_ available in _OE-Core_, but in this case I would rather set the tune to `cortexa53-crypto` instead.

File: recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    # Synopsys DDR
+    for ddr_firmware in ${DDR_FIRMWARE_NAME}; do
+        install -m 0644 ${S}/firmware/ddr/synopsys/${ddr_firmware} ${DEPLOYDIR}
+    done
+}
+addtask deploy after do_install before do_build
+
+PACKAGE_ARCH = "${MACHINE_SOCARCH}"
+
+COMPATIBLE_MACHINE = "(mx9-nxp-bsp)"
```
Comment (zandrey) [inline recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb:21]: Why is this file required? It is a copy of `firmware-imx-8m`, which can be used in this context as well.

File: conf/machine/imx93evk.conf [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#@TYPE: Machine
```
Comment (thochstein) [inline conf/machine/imx93evk.conf:1]: This file is used by NXP for internal testing and shouldn't be upstreamed.

File: conf/machine/include/imx-base.inc (hunk 1 above)
Comment (thochstein) [reply conf/machine/include/imx-base.inc:167]: Looks like A55 is available?

https://github.com/openembedded/openembedded-core/blob/master/meta/conf/machine/include/arm/armv8-2a/tune-cortexa55.inc

File: recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig [hunk 4]
```diff
@@ -0,0 +1,1071 @@
+CONFIG_SYSVIPC=y
```
Comment (thochstein) [inline recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig:1]: What is the purpose of the default `defconfig` for all `mx9`? At the least the commit should explain why we add it, but perhaps even better would be to not add it until it is actually used?

File: conf/machine/include/imx-base.inc (hunk 1 above)
Comment (zandrey) [reply conf/machine/include/imx-base.inc:167]: True, thanks for the hint!

Sorry, I was searching in the different folder (`armv8a`) and did not realize that _OE-Core_ had a separate folder laid out for _ARMv8.2-A_ architecture.

Then it's indeed not correct - the tune shall be set to point to the correct one.

File: classes/use-imx-security-controller-firmware.bbclass [hunk 5]
```diff
@@ -24,6 +24,7 @@ SECO_FIRMWARE_NAME:mx8qxp-nxp-bsp = \
 SECO_FIRMWARE_NAME:mx8dx-nxp-bsp  = "mx8qxc0-ahab-container.img"
 SECO_FIRMWARE_NAME:mx8dxl-nxp-bsp = "mx8dxla1-ahab-container.img"
 SECO_FIRMWARE_NAME:mx8ulp-nxp-bsp = "mx8ulpa1-ahab-container.img"
+SECO_FIRMWARE_NAME:mx9-nxp-bsp    = "mx93a0-ahab-container.img"
```
Comment (otavio) [inline classes/use-imx-security-controller-firmware.bbclass:27]: Please, use `mx93` as override.

File: conf/machine/include/imx-base.inc [hunk 6]
```diff
@@ -372,6 +379,7 @@ IMX_EXTRA_FIRMWARE:mx8-generic-bsp   = "firmware-imx-8 imx-sc-firmware imx-seco"
 IMX_EXTRA_FIRMWARE:mx8m-generic-bsp  = "firmware-imx-8m"
 IMX_EXTRA_FIRMWARE:mx8x-generic-bsp  = "imx-sc-firmware imx-seco"
 IMX_EXTRA_FIRMWARE:mx8ulp-generic-bsp = "firmware-upower firmware-sentinel"
+IMX_EXTRA_FIRMWARE:mx9-nxp-bsp        = " firmware-imx-8 firmware-sentinel"
```
Comment (otavio) [inline conf/machine/include/imx-base.inc:382]: Please, use `mx93` override.

File: recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb (hunk 2 above)
Comment (otavio) [reply recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb:21]: This file could be removed, right?

File: recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig (hunk 4 above)
Comment (otavio) [reply recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig:1]: I'd let the `mx93` one, only.

**Review by otavio — CHANGES_REQUESTED** [review]

File: recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig (hunk 4 above)
Comment (angolini) [reply recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig:1]: I will fix the commit log, but they are the same defconfig, but one for `linux-imx`and one for `linux-fslc-imx`. It happens that today they are the same.
Is there a better way to have one file only for both linux?

File: conf/machine/include/imx-base.inc (hunk 1 above)
Comment (angolini) [reply conf/machine/include/imx-base.inc]: changed

File: recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig (hunk 4 above)
Comment (thochstein) [reply recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig:1]: Ah, I did not notice the two files are for different linux. For `linux-imx`, I see that this is the same `defconfig` as for all i.MX 8. Perhaps we can link `linux-imx/imx9-nxp-bsp/defconfig` to `linux-imx/imx8-nxp-bsp/defconfig`? Or better, can we replace both with `linux-imx/aarch64/defconfig`?

Comment (otavio) [reply recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig:1]: Using `aarch64` seems like a good option.

File: conf/machine/include/imx-base.inc (hunk 6 above)
Comment (thochstein) [reply conf/machine/include/imx-base.inc:382]: FYI, I started to suggest that we use `mx9` to denote the family as we do for `mx8m` and `mx8x`, but I remembered that the naming scheme has changed and `mx8m` is analogous to `mx93` or `mx95`. This is an improvement on the naming since for i.MX 8 we had the ambiguity of an i.MX 8 Series and an i.MX 8 Family.

File: recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb (hunk 2 above)
Comment (angolini) [reply recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb:21]: I'm currently testing this

File: conf/machine/include/imx-base.inc (hunk 1 above)
Comment (zandrey) [reply conf/machine/include/imx-base.inc]: @angolini 

There is care to be taken here, as changing `DEFAULTTUNE` to plain `cortexa55` here would not imply that Advanced SIMD instructions would be enabled as well... This was enabled implicitly by `-crypto` suffix for `imx8m` series, but as far as I can see this suffix is not (yet) supported in _OE_Core_ for _A55_ profile.

I'm not sure if it's a miss in _OE-Core_ or it is not brought up deliberately, but without it - performance might suffer as ASIMD instructions would not be emitted by compiler.

@thochstein @otavio I'm a bit torn here: from one side - the new tune matches to what SOC has in the Cortex-A complex, but on the other hand - resulting compiled binaries might suffer performance or size difference if it is enabled like this.

Shall it be taken as it is presented in this PR, or perhaps first `-crypto` suffix shall be submitted to _OE-Core_?

I'm not sure: if `imx93` implements full _A55_ profile(@thochstein ❔) - then `-crypto` suffix would really be desired; if not - then this change is good to go.

-- andrey

Comment (thochstein) [reply conf/machine/include/imx-base.inc]: I will have to ask. We have been setting `DEFAULTTUNE:mx9-nxp-bsp    ?= "armv8a-crc-crypto"` in our distros, and I see now that we neglected to add an optimized tuning from the BSP layer.

Comment (otavio) [reply conf/machine/include/imx-base.inc]: I'd add it for our tune, so it uses by default @zandrey.

Comment (angolini) [reply conf/machine/include/imx-base.inc]: so far, I will let it be. Waiting to see if @thochstein brings some news

File: recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb (hunk 2 above)
Comment (angolini) [reply recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb:21]: I got it working somehow, I will update the code so you can review. After review I can test other SoCs

File: conf/machine/include/imx-base.inc (hunk 6 above)
Comment (junzhuimx) [inline conf/machine/include/imx-base.inc:382]: firmware-imx-9 contains for DDR firmware for i.MX93. 
it is not right to use firmware-imx-8

File: recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb (hunk 2 above)
Comment (thochstein) [reply recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb:21]: @otavio, wouldn't it be better to combine all these different `firmware-imx` recipes since they are all using the same `SRC_URI`?

Comment (zandrey) [reply recipes-bsp/firmware-imx/firmware-imx-9_8.18.bb:21]: > @zandrey it is no the copy of firmware-imx-8m. recently, firmware-imx-9 is for i.MX 9 series firmware only.


From the recipe content perspective, it is a copy.

I tend to agree with @thochstein comment above to combine `firmware-imx` into one recipe, since `8m` and `9` in this context point to the same `SRC_URI`.

File: conf/machine/include/imx-base.inc (hunk 1 above)
Comment (zandrey) [reply conf/machine/include/imx-base.inc]: @thochstein

Good catch! Yes, indeed it is!

It is actually coming from the top include, and shall be enabled for `cortexa55`.

I would suggest to set `DEFAULTTUNE` straight to `cortexa55` here.

Comment (thochstein) [reply conf/machine/include/imx-base.inc]: @zandrey, great! I've also received confirmation that i.MX 93 implements full A55 profile.

Comment (angolini) [reply conf/machine/include/imx-base.inc]: great! So I'm resolving this as we all agreed (and my current version is already fixed)

File: recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig (hunk 4 above)
Comment (angolini) [reply recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig:1]: then using `aarch64` the `imx-nxp-bsp` is avoided which causes an error for all the machines, so I suggest we use `imx8-nxp-bsp` and `imx9-nxp-bsp`

File: conf/machine/include/imx-base.inc (hunk 6 above)
Comment (angolini) [reply conf/machine/include/imx-base.inc:382]: I have a patch on that recipe to install the right files. Let me push my latest

File: recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig (hunk 4 above)
Comment (thochstein) [reply recipes-kernel/linux/linux-fslc-imx/mx9-nxp-bsp/defconfig:1]: Can we change `imx-nxp-bsp` to `arm`?

File: recipes-bsp/firmware-imx/firmware-imx-8_8.18.bb [hunk 7]
```diff
@@ -18,6 +18,6 @@ addtask deploy after do_install before do_build
 
 PACKAGE_ARCH = "${MACHINE_SOCARCH}"
 
-COMPATIBLE_MACHINE      = "(mx8-nxp-bsp)"
+COMPATIBLE_MACHINE      = "(mx8-nxp-bsp|mx9-nxp-bsp)"
```
Comment (thochstein) [inline recipes-bsp/firmware-imx/firmware-imx-8_8.18.bb:21]: With `firmware-imx-9` back, this change needs to be dropped as well.

**Comment by angolini** [comment]
@otavio @thochstein @zandrey 
The changes are covered. The missing points are:

* for linux-imx/linux-fslc-imx what should be the directory structure?
currently we have (after this PR):
```
recipes-kernel/linux/linux-fslc
├── armv8a
├── defconfig
└── mxs-generic-bsp
recipes-kernel/linux/linux-fslc-imx
├── imx-nxp-bsp
├── mx8-nxp-bsp
└── mx93-nxp-bsp
recipes-kernel/linux/linux-fslc-lts
├── armv8a
├── defconfig
└── mxs-generic-bsp
recipes-kernel/linux/linux-imx
├── imx-nxp-bsp
├── mx8-nxp-bsp
└── mx93-nxp-bsp
```
* firmware-imx* should be merged?
currently after this PR we have:
```
./recipes-bsp/firmware-imx/firmware-imx-8_8.18.bb
./recipes-bsp/firmware-imx/firmware-imx_8.18.bb
./recipes-bsp/firmware-imx/firmware-imx-8m_8.18.bb
./recipes-bsp/firmware-imx/firmware-imx-8.18.inc
```

File: recipes-bsp/firmware-imx/firmware-imx-8_8.18.bb (hunk 7 above)
Comment (junzhuimx) [reply recipes-bsp/firmware-imx/firmware-imx-8_8.18.bb:37]: @angolini @otavio 
firmware-imx-9 is integrated to firmware-imx-8 recipe. 
How about also merge firmware-imx-8m into formware-imx-8?

Comment (otavio) [reply recipes-bsp/firmware-imx/firmware-imx-8_8.18.bb:37]: @junzhuimx, totally agree. See #1426.

**Comment by angolini** [comment]
From my POV all the requests were covered (or are tracked in an issue)

PR updated with the latest working version

**Review by otavio — APPROVED** [review]

**Review by thochstein — APPROVED** [review]
